pkg/controller/podautoscaler/horizontal_test.go

```diff
@@ -1720,4 +1720,89 @@ func TestAvoidUncessaryUpdates(t *testing.T) {
 	tc.runTestWithController(t, controller, informerFactory)
 }
 
+func TestConvertDesiredReplicasWithRules(t *testing.T) {
+	conversionTestCases := []struct {
+		currentReplicas                  int32
+		desiredReplicas                  int32
+		hpaMinReplicas                   int32
+		hpaMaxReplicas                   int32
+		expectedConvertedDesiredReplicas int32
+		annotation                       string
+	}{
+		{
+			currentReplicas:                  5,
+			desiredReplicas:                  7,
+			hpaMinReplicas:                   3,
+			hpaMaxReplicas:                   8,
+			expectedConvertedDesiredReplicas: 7,
+			annotation:                       "prenormalized desired replicas within range",
+		},
+		{
+			currentReplicas:                  3,
+			desiredReplicas:                  1,
+			hpaMinReplicas:                   2,
+			hpaMaxReplicas:                   8,
+			expectedConvertedDesiredReplicas: 2,
+			annotation:                       "prenormalized desired replicas < minReplicas",
+		},
+		{
+			currentReplicas:                  1,
+			desiredReplicas:                  0,
+			hpaMinReplicas:                   0,
+			hpaMaxReplicas:                   10,
+			expectedConvertedDesiredReplicas: 1,
+			annotation:                       "1 is minLimit because hpaMinReplicas < 1",
+		},
+		{
+			currentReplicas:                  20,
+			desiredReplicas:                  1000,
+			hpaMinReplicas:                   1,
+			hpaMaxReplicas:                   10,
+			expectedConvertedDesiredReplicas: 10,
+			annotation:                       "maxReplicas is the limit because maxReplicas < scaleUpLimit",
+		},
+		{
+			currentReplicas:                  3,
+			desiredReplicas:                  1000,
+			hpaMinReplicas:                   1,
+			hpaMaxReplicas:                   2000,
+			expectedConvertedDesiredReplicas: calculateScaleUpLimit(3),
+			annotation:                       "scaleUpLimit is the limit because scaleUpLimit < maxReplicas",
+		},
```

Comment:
Can we have some tests with currentReplicas == 0? I know it technically can't happen right now as we will go to a different path in reconcileAutoscaler, but there is no reason why we shouldn't be able to handle it here.

Reply:
Sure! Will add.

Reply:
currentReplicas == 0 explicitly disables the HPA, FYI

Reply:
reconcileAutoscaler has an explicit if to deal with 0 replicas: https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/podautoscaler/horizontal.go#L404. Or is it going to disable even before getting to reconcileAutoscaler?

Reply:
I don't plan on removing the part in L404 that deals with the 0 condition (at least not yet). I envisioned the code I'm writing now as handling the normalization of a prediction made using metrics. In that case, it would allow a different part of the execution path to handle when `currentReplicas == 0`.